# COMP20230 - Assignment 2

In [7]:
from scipy.optimize import curve_fit
import timeit
import matplotlib.pyplot as plt
import numpy as np
import random
import math

## Section 1 - Sorting Algorithms

### Section 1.1 - Implement Sorting Algorithms

#### Bubble Sort

In [9]:
def bubble_sort(*args):
    """Implementation of the Bubble Sort Algorithm
    
    For each index in the array, comparison operations are performed to push the largest element to the top of the array
    Once the element is at the top of the array, we forget about it, and move onto the previous position in the array
    We do this until no more comparison operations have been performed - indicating that the array has been sorted
    Reference: https://stackabuse.com/bubble-sort-in-python/
    """
    our_list = args[0]

    has_swapped = True

    num_of_iterations = 0

    while(has_swapped):
        has_swapped = False
        for i in range(len(our_list) - num_of_iterations - 1):
            if our_list[i] > our_list[i+1]:
                # Swap
                our_list[i], our_list[i+1] = our_list[i+1], our_list[i]
                has_swapped = True
        num_of_iterations += 1

    return our_list

#### Quick Sort

In [10]:
def quick_sort(*args):
    """Implementation of the Quick Sort Algorithm
    
    Reference: COMP20230 lab material
    """
    data = args[0]

    if len(data) > 1:
        less = []
        equal = []
        greater = []
        pivot = data[0]

        for x in data:
            if x < pivot:
                less.append(x)
            elif x == pivot:
                equal.append(x)
            else:
                greater.append(x)

        return quick_sort(less) + equal + quick_sort(greater)

    else:
        return data

#### Merge Sort

In [11]:
def merge(left, right):
    result = []
    left_idx, right_idx = 0, 0
    while left_idx < len(left) and right_idx < len(right):
        if left[left_idx] <= right[right_idx]:
            result.append(left[left_idx])
            left_idx += 1
        else:
            result.append(right[right_idx])
            right_idx += 1
 
    if left:
        result.extend(left[left_idx:])
    if right:
        result.extend(right[right_idx:])
    return result


def mergesort(*args):
    """Implementation of the Merge Sort Algorithm
    
    Reference: COMP20230 lab material
    """
    arr = args[0]

    if len(arr) <= 1:
        return arr
 
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]
 
    left = mergesort(left)
    right = mergesort(right)
    return list(merge(left, right))

In [12]:
test_arr = [4, 3, 2, 1]
print(bubble_sort(test_arr))
print(quick_sort(test_arr))
print(mergesort(test_arr))

[1, 2, 3, 4]
[1, 2, 3, 4]
[1, 2, 3, 4]


### Section 1.2: Create the input lists

### Section 1.3: Create helper functions

#### Create function to calculate runtimes

In [8]:
def get_runtimes(*args, funct=bubble_sort):
    """Function that takes in a series of positional arguments and a function and returns the runtime of function(arguments)"""
        
    # Compute the runtime of the algorithm using the timeit module
    start = timeit.default_timer()
    funct(*args)
    stop = timeit.default_timer()
    runtime = stop - start

    return runtime

In [13]:
print(get_runtimes(test_arr, funct=bubble_sort))
print(get_runtimes(test_arr, funct=quick_sort))
print(get_runtimes(test_arr, funct=mergesort))

8.399991202168167e-06
1.3900003978051245e-05
2.1100000594742596e-05


### Section 1.4: Calculate the sorting algorithm runtimes

### Section 1.5: Discussion of results